### PR TITLE
Add initial headless mode support

### PR DESCRIPTION
**Pull request by ThomasJClark**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -60,9 +60,18 @@ project(":core") {
...
+            dirs 'libs'
+        }
+    }
+
     dependencies {
         compile group: 'org.bytedeco', name: 'javacv', version: '1.1'
         compile group: 'org.bytedeco.javacpp-presets', name: 'opencv', version: '3.0.0-1.1'
+        compile group: 'org.bytedeco.javacpp-presets', name: 'opencv', version: '3.0.0-1.1', classifier: os
+        compile group: 'org.bytedeco.javacpp-presets', name: 'opencv-3.0.0-1.1', classifier: "linux-frc"
+        compile group: 'org.bytedeco.javacpp-presets', name: 'videoinput', version: '0.200-1.1', classifier: os
```
Comment (JLLeitschuh) [inline build.gradle]: Add a flag

Comment (JLLeitschuh) [inline build.gradle]:
```
if (project.hasProperty('coreOnly')) {
  compile group: 'org.bytedeco.javacpp-presets', name: 'videoinput', version: '0.200-1.1', classifier: os
} else {
  testCompile group: 'org.bytedeco.javacpp-presets', name: 'videoinput', version: '0.200-1.1', classifier: os
}
```

File: core/src/main/java/edu/wpi/grip/core/Main.java [hunk 2]
```diff
@@ -1,6 +1,38 @@
...
-    public static void main(String[] args) {
+    public static void main(String[] args) throws Exception {
+        if (args.length != 1) {
+            System.err.println("Usage: GRIP.jar project.grip");
+            return;
+        }
+
+        final String projectPath = args[0];
+
+        final EventBus eventBus = new EventBus();
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/Main.java]: This should at least print the stack trace. The built in logger for the event bus sucks.

File: core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
+/*
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java]: This class isn't even used anywhere?

Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java]: Also, IPCameraFrameGrabber isn't a source, It should probably go in a different package.

File: core/src/main/java/edu/wpi/grip/core/Main.java [hunk 4]
```diff
@@ -1,6 +1,38 @@
...
+        final Pipeline pipeline = new Pipeline(eventBus);
+        final Palette palette = new Palette(eventBus);
+        final Project project = new Project(eventBus, pipeline, palette);
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/Main.java]: We should have one way of creating these objects and the UI should take advantage of it.
Do something like `main.setup()`.
That way if we change any of the constructors we don't need to worry about it.

As for dependency injection, lets take care of that when it actually gets implemented.

File: core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java (hunk 3 above)
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java]: Can you add very clear comments on the lines you've added/changed to be completely clear what is different.
Do it inline like `// Our change`

File: core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        BufferedImage bi = ImageIO.read(new ByteArrayInputStream(readImage()));
+        return bi;
+    }
+
+    byte[] readImage() throws IOException {
+        byte[] buffer = new byte[4096];// MTU or JPG Frame Size?
+        int n = -1;
+
+        StringBuffer sb = new StringBuffer();
+        int total = 0;
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java]: This variable is no longer used.

File: core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java (hunk 3 above)
Comment (ThomasJClark) [inline core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java]: > This class isn't even used anywhere?

That's not true

File: core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    public BufferedImage grabBufferedImage() throws IOException {
+        BufferedImage bi = ImageIO.read(new ByteArrayInputStream(readImage()));
+        return bi;
+    }
+
+    byte[] readImage() throws IOException {
+        byte[] buffer = new byte[4096];// MTU or JPG Frame Size?
+        int n = -1;
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java]: This variable is no longer used.

File: core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+        }
+    }
+
+    public BufferedImage grabBufferedImage() throws IOException {
+        BufferedImage bi = ImageIO.read(new ByteArrayInputStream(readImage()));
+        return bi;
+    }
+
+    byte[] readImage() throws IOException {
+        byte[] buffer = new byte[4096];// MTU or JPG Frame Size?
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java]: This variable is no longer used.

File: core/src/main/java/edu/wpi/grip/core/Main.java (hunk 4 above)
Comment (ThomasJClark) [inline core/src/main/java/edu/wpi/grip/core/Main.java]: > As for dependency injection, lets take care of that when it actually gets implemented.
 Add a line note

This is the answer to the first part of your comment

File: build.gradle [hunk 8]
```diff
@@ -60,15 +60,22 @@ project(":core") {
...
+            dirs 'libs'
+        }
+    }
+
     dependencies {
         compile group: 'org.bytedeco', name: 'javacv', version: '1.1'
         compile group: 'org.bytedeco.javacpp-presets', name: 'opencv', version: '3.0.0-1.1'
+        compile group: 'org.bytedeco.javacpp-presets', name: 'opencv', version: '3.0.0-1.1', classifier: os
+        compile group: 'org.bytedeco.javacpp-presets', name: 'opencv-3.0.0-1.1', classifier: 'linux-frc'
+        compile group: 'org.bytedeco.javacpp-presets', name: 'videoinput', version: '0.200-1.1', classifier: os
```
Comment (JLLeitschuh) [inline build.gradle]: Still need to fix this.

Comment (ThomasJClark) [inline build.gradle]: :-1: this works fine and doesn't have a huge impact on file size

File: core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java (hunk 3 above)
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java]: I don't see any changes in the CameraSource. How is this one being imported? Doesn't the bytdeco import take precedence?

Comment (ThomasJClark) [inline core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java]: no

Comment (ThomasJClark) [inline core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java]: this one does

File: core/src/main/java/edu/wpi/grip/core/Main.java [hunk 9]
```diff
@@ -1,6 +1,38 @@
...
+        final EventBus eventBus = new EventBus((exception, context) -> exception.printStackTrace());
+        final Pipeline pipeline = new Pipeline(eventBus);
+        final Palette palette = new Palette(eventBus);
+        final Project project = new Project(eventBus, pipeline, palette);
+
+        Operations.addOperations(eventBus);
+        CVOperations.addOperations(eventBus);
+
+        // Open a project from a .grip file specified on the command line
+        project.open(new File(projectPath));
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/Main.java]: Either add dependency injection first or make one way to construct GRIP.
It should probably be wrapped an object like `GripInstance`

**Pull request merged**